src/utility_stmt_hooking.c

```diff
@@ -689,20 +704,78 @@ PathmanCopyFrom(CopyState cstate, Relation parent_rel,
 }
 
 /*
- * COPY FROM does not support FDWs, emit ERROR.
+ * Init COPY FROM, if supported.
  */
 static void
-prepare_rri_for_copy(EState *estate,
-					 ResultRelInfoHolder *rri_holder,
-					 const ResultPartsStorage *rps_storage,
-					 void *arg)
+prepare_rri_for_copy(ResultRelInfoHolder *rri_holder,
+					 const ResultPartsStorage *rps_storage)
 {
-	ResultRelInfo  *rri = rri_holder->result_rel_info;
-	FdwRoutine	   *fdw_routine = rri->ri_FdwRoutine;
+	ResultRelInfo	*rri = rri_holder->result_rel_info;
+	FdwRoutine		*fdw_routine = rri->ri_FdwRoutine;
 
 	if (fdw_routine != NULL)
+	{
+		/*
+		 * If this Postgres has no idea about shardman, behave as usual:
+		 * vanilla Postgres doesn't support COPY FROM to foreign partitions.
+		 * However, shardman patches to core extend FDW API to allow it,
+		 * though currently postgres_fdw does so in a bit perverted way: we
+		 * redirect COPY FROM to parent table on foreign server, assuming it
+		 * exists, and let it direct tuple to proper partition. This is
+		 * because otherwise we have to modify logic of managing connections
+		 * in postgres_fdw and keep many connections open to one server from
+		 * one backend.
+		 */
+#ifndef PG_SHARDMAN
+		goto bail_out; /* to avoid 'unused label' warning */
+#else
+		{ /* separate block to avoid 'unused var' warnings */
+			CopyState		cstate = (CopyState) rps_storage->callback_arg;
+			ResultRelInfo	*parent_rri;
+			const char		*parent_relname;
+			EState		   	*estate;
+
+			/* shardman COPY FROM requested? */
+			if (*find_rendezvous_variable(
+					"shardman_pathman_copy_from_rendezvous") == NULL)
+				goto bail_out;
+
+			estate = rps_storage->estate;
+			parent_rri = rps_storage->saved_rel_info;
+			parent_relname = psprintf(
+				"%s.%s", "public",
+				quote_identifier(RelationGetRelationName(parent_rri->ri_RelationDesc)));
+			if (!FdwCopyFromIsSupported(fdw_routine))
+				ereport(ERROR,
+						(errcode(ERRCODE_WRONG_OBJECT_TYPE),
+						 errmsg("FDW adapter for relation \"%s\" doesn't support COPY FROM",
+								RelationGetRelationName(rri->ri_RelationDesc))));
+			fdw_routine->BeginForeignCopyFrom(estate, rri, cstate, parent_relname);
+			return;
+		}
+#endif
+bail_out:
 		elog(ERROR, "cannot copy to foreign partition \"%s\"",
 			 get_rel_name(RelationGetRelid(rri->ri_RelationDesc)));
+	}
+}
+
+/*
+ * Shut down FDWs.
+ */
+static void
+finish_rri_copy(ResultRelInfoHolder *rri_holder,
```

Comment:
I think this should be passed as a "delete callback". Maybe we should call them `rri_init_callback` & `rri_fini_callback`?